File: multitypelistviewadapter/src/main/java/net/kmfish/multitypelistviewadapter/BaseListAdapter.java
package net.kmfish.multitypelistviewadapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import java.util.List;

/**
 * Created by kmfish on 2015/9/9
 */
public class BaseListAdapter extends BaseAdapter implements NotifyDataSetChangedListener, IArrayAdapter {

    public static final String TAG = BaseListAdapter.class.getSimpleName();

    private BaseArrayAdapter arrayAdapter;

    private LayoutInflater mInflater;

    private ListItemFactory itemFactory = new ClassListItemFactory();

    public BaseListAdapter() {
        init();
    }

    private void init() {
        arrayAdapter = new BaseArrayAdapter(this);
    }

    /**
     * 初始化Adapter时调用此方法注册数据类型和Item类型的映射关系
     * 注意: 需要在setAdapter之前调用!!
     * @param dataClz model class
     * @param itemClz ListItem class
     * @param <ModelType> model类型
     * @param <AttachType> 附加对象类型
     */
    public <ModelType, AttachType> void registerDataAndItem(Class<ModelType> dataClz,
                                                            Class<? extends ListItem<ModelType, AttachType>> itemClz) {
        itemFactory.registerDataType(dataClz, itemClz);
    }

    /**
     * 初始化Adapter时调用此方法注册数据类型和Item类型的映射关系
     * 注意: 需要在setAdapter之前调用!!
     * @param dataClz model class
     * @param itemClz ListItem class
     * @param attachInfo 附加对象
     * @param <ModelType> model类型
     * @param <AttachType> 附加对象类型
     */
    public <ModelType, AttachType> void registerDataAndItem(Class<ModelType> dataClz,
                                                            Class<? extends ListItem<ModelType, AttachType>> itemClz,
                                                            AttachType attachInfo) {
        itemFactory.registerDataType(dataClz, itemClz);
        itemFactory.registerDataClickListener(dataClz, attachInfo);
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if (null == mInflater) {
            mInflater = LayoutInflater.from(parent.getContext());
        }

        Data data = getData(position);
        if (null == data || null == data.getData()) {
            throw new RuntimeException("item data should not be null, pos:" + position);
        }

        Class rawClz = data.getData().getClass();
        ListItem item;
        if (null == convertView) {
            item = createListItem(rawClz);
            if (null != item) {
                convertView = mInflater.inflate(item.onGetLayoutRes(), parent, false);
                convertView.setTag(item);
                item.setContext(parent.getContext());
                item.bindViews(convertView);
            }
        } else {
            item = (ListItem) convertView.getTag();
        }

        if (null != item) {
            item.setData(data.getData());
            item.updateView(data.getData(), position);
        }
        return convertView;
    }

    private ListItem createListItem(Class<?> dataClz) {
        ListItem item = itemFactory.create(dataClz);
        return item;
    }

    @Override
    public boolean isEnabled(int position) {
        Data item = getItem(position);
        if (null != item) {
            return item.isEnable();
        }

        return super.isEnabled(position);
    }

    @Override
    public int getViewTypeCount() {
        int count = itemFactory.getTypeCount();
        if (0 == count) {
            throw new RuntimeException(
                    "registerDataAndItem() should be called before ListView.setAdapter(adapter)!");
        }

        return itemFactory.getTypeCount();
    }

    @Override
    public int getItemViewType(int position) {
        return arrayAdapter.getItemType(position);
    }

    @Override
    public int getCount() {
        return arrayAdapter.getCount();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public Data getItem(int position) {
        return arrayAdapter.getData(position);
    }

    @Override
    public Data getData(int position) {
        return arrayAdapter.getData(position);
    }

    @Override
    public void addData(Data item) {
        arrayAdapter.addData(item);
    }

    @Override
    public void addDatas(List<Data> items) {
        arrayAdapter.addDatas(items);
    }

    @Override
    public void addDatas(Data... items) {
        arrayAdapter.addDatas(items);
    }

    @Override
    public void insert(Data item, int index) {
        arrayAdapter.insert(item, index);
    }

    @Override
    public void remove(Data object) {
        arrayAdapter.remove(object);
    }

    @Override
    public void remove(int pos) {
        arrayAdapter.remove(pos);
    }

    @Override
    public void clear() {
        arrayAdapter.clear();
    }

    @Override
    public void setNotifyOnChange(boolean notifyOnChange) {
        arrayAdapter.setNotifyOnChange(notifyOnChange);
    }
}

File: multitypelistviewadapter/src/main/java/net/kmfish/multitypelistviewadapter/BaseListAdapter.kt
package net.kmfish.multitypelistviewadapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter

/**
 * Created by kmfish on 2015/9/9
 */
class BaseListAdapter : BaseAdapter(), NotifyDataSetChangedListener, IArrayAdapter {

    private lateinit var arrayAdapter: BaseArrayAdapter

    private var mInflater: LayoutInflater? = null

    private val itemFactory = ClassListItemFactory()

    init {
        init()
    }

    private fun init() {
        arrayAdapter = BaseArrayAdapter(this)
    }

    /**
     * 初始化Adapter时调用此方法注册数据类型和Item类型的映射关系
     * 注意: 需要在setAdapter之前调用!!
     * @param dataClz model class
     * *
     * @param itemClz ListItem class
     * *
     * @param <ModelType> model类型
     * *
     * @param <AttachType> 附加对象类型
    </AttachType></ModelType> */
    fun <ModelType, AttachType : Any> registerDataAndItem(dataClz: Class<ModelType>,
                                                    itemClz: Class<out ListItem<ModelType, AttachType>>) {
        itemFactory.registerDataType(dataClz, itemClz)
    }

    /**
     * 初始化Adapter时调用此方法注册数据类型和Item类型的映射关系
     * 注意: 需要在setAdapter之前调用!!
     * @param dataClz model class
     * *
     * @param itemClz ListItem class
     * *
     * @param attachInfo 附加对象
     * *
     * @param <ModelType> model类型
     * *
     * @param <AttachType> 附加对象类型
    </AttachType></ModelType> */
    fun <ModelType, AttachType : Any> registerDataAndItem(dataClz: Class<ModelType>,
                                                    itemClz: Class<out ListItem<ModelType, AttachType>>,
                                                    attachInfo: AttachType) {
        itemFactory.registerDataType(dataClz, itemClz)
        itemFactory.registerDataClickListener(dataClz, attachInfo)
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {
        var cvtView = convertView
        if (null == mInflater) {
            mInflater = LayoutInflater.from(parent.context)
        }

        val data = getData(position)
        if (null == data || null == data.data) {
            throw RuntimeException("item data should not be null, pos:" + position)
        }

        val dataInner = data.data
        val rawClz = dataInner.javaClass
        if (null == cvtView) {
            val item = createListItem(rawClz)
            if (null != item) {
                cvtView = mInflater!!.inflate(item.onGetLayoutRes(), parent, false)
                cvtView!!.tag = item
                item.setContext(parent.context)
                item.bindViews(cvtView)

                item.data = dataInner
                item.updateView(dataInner, position)
            }
        } else {
            val item = castListItem(cvtView.tag, rawClz)
            item.data = dataInner
            item.updateView(dataInner, position)
        }

        return cvtView
    }

    private fun <T> castListItem(viewTag: Any?, dataClz: Class<T>): ListItem<T, *> {
        val item = viewTag as ListItem<T, *>
        return item
    }

    private fun <T> createListItem(dataClz: Class<T>): ListItem<T, *>? {
        return itemFactory.create(dataClz)
    }

    override fun isEnabled(position: Int): Boolean {
        val item = getItem(position)
        if (null != item) {
            return item.isEnable
        }

        return super.isEnabled(position)
    }

    override fun getViewTypeCount(): Int {
        val count = itemFactory.typeCount
        if (0 == count) {
            throw RuntimeException(
                    "registerDataAndItem() should be called before ListView.setAdapter(adapter)!")
        }

        return itemFactory.typeCount
    }

    override fun getItemViewType(position: Int): Int {
        return arrayAdapter.getItemType(position)
    }

    override fun getCount(): Int {
        return arrayAdapter.count
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItem(position: Int): Data<*>? {
        return arrayAdapter.getData(position)
    }

    override fun getData(position: Int): Data<*>? {
        return arrayAdapter.getData(position)
    }

    override fun addData(item: Data<*>) {
        arrayAdapter.addData(item)
    }

    override fun addDatas(items: List<Data<*>>) {
        arrayAdapter.addDatas(items)
    }

    override fun addDatas(vararg items: Data<*>) {
        arrayAdapter.addDatas(*items)
    }

    override fun insert(item: Data<*>, index: Int) {
        arrayAdapter.insert(item, index)
    }

    override fun remove(`object`: Data<*>) {
        arrayAdapter.remove(`object`)
    }

    override fun remove(pos: Int) {
        arrayAdapter.remove(pos)
    }

    override fun clear() {
        arrayAdapter.clear()
    }

    override fun setNotifyOnChange(notifyOnChange: Boolean) {
        arrayAdapter.setNotifyOnChange(notifyOnChange)
    }

    companion object {

        val TAG = BaseListAdapter::class.java.simpleName
    }
}